File: leetcode/c++/0153-find-minimum-in-rotated-sorted-array.cc
// https://leetcode.com/problems/find-minimum-in-rotated-sorted-array/
// Medium

#include<vector>
using namespace std;
class Solution {
public:
    int findMin(vector<int>& nums) {
        int lo = 0, hi = nums.size()-1;
        int mid;
        while (lo < hi) {
            mid = (lo + hi) / 2;
            //cout << lo << ' ' << mid << ' ' << hi << endl;
            if (nums[lo] < nums[hi]) return nums[lo];
            else if (mid == lo) lo = mid + 1;
            else if (mid == hi) hi = mid - 1;
            else if (nums[mid] > nums[lo]) {
                lo = mid;
            } else {
                hi = mid;
            }
        }
        return nums[lo];
    }
};

File: leetcode/python/0153-find-minimum-in-rotated-sorted-array.py
# https://leetcode.com/problems/find-minimum-in-rotated-sorted-array/
# Medium

# Perform a binary search with lo and hi
# If nums[mid-1] >= nums[mid] <= nums[mid+1] then you found min
# Otherwise compare nums[mid] to nums[hi] and move lo and hi accordingly
class Solution:
    def findMin(self, nums: List[int]) -> int:
        lo, hi = 0, len(nums)-1
        while lo <= hi:
            mid = (lo + hi) // 2
            # ----Alternative IF statement which is cleaner but harder to grasp---
            # if nums[mid] < nums[hi]:
            #   hi = mid
            # else:
            #   lo = mid + 1

            if nums[mid-1] >= nums[mid] <= nums[(mid+1)%len(nums)]:
                return nums[mid]
            elif nums[mid] < nums[hi]:
                hi = mid - 1
            else:
                lo = mid + 1
        return -5001
                